### ISSUE TITLE
New check: TestMethodVisibility

### ISSUE THREAD

**Issue by Michael1993** [checkstyle/checkstyle#7176]
As of JUnit 5, test methods don't have to be public.
https://junit.org/junit5/docs/snapshot/user-guide/#writing-tests-classes-and-methods

> "Test classes, test methods, and lifecycle methods are not required to be public, but they must not be private"

According to Item 15 of Joshua Bloch Effective Java, Third Edition (Minimize the accessibility of classes and members) a programmer should:
> "[...] make each class or member as inaccessible as possible."

So I suggest a new check that checks that **all** methods annotated with `@Test` in a project has the same visibility, by default package-private (This is also in compliance with the FIRST principles of Unit testing).

**Comment by romani** [comment]
>This is also in compliance with the FIRST principles of Unit testing

Please do a quote, with link to source

**Comment by romani** [comment]
> project has the same visibility

I presume you need Check that is unrelated to junit at all, as junit might change something else, and will never  be up-to-date.
What if future Check just validate match of visibility scope if some annotation is placed on method ?

**Comment by Michael1993** [comment]
That sounds perfect if I can specify both the annotation and the required visibility.

**Comment by romani** [comment]
To approve issue I need to have ready to development design, in this case issue have good chance to be implemented. Are you ready to contribute exact design (names of Check and properties, expected behavior)? I will guide.

**Comment by Michael1993** [comment]
Yes, of course.

**Comment by romani** [comment]
>I thought I'd have to implement it myself?

This is the quickest way to get what you want.

Before any coding start let's agree on design and behavior. Please review good example of what we need for design proposal
https://github.com/checkstyle/checkstyle/issues/5884 .  Please update your issue description to have kind of same details.

**Comment by Michael1993** [comment]
@romani Please review.

**Comment by verhas** [comment]
...
  
  <property name="annotation" value="org.junit.jupiter.api.Test"/>
...

**Comment by romani** [comment]
Should property be "annotation" single class name or "annotations" array of names ? With singular property, user will have to create multiple instances of Check for each annotation.
According to https://junit.org/junit5/docs/current/user-guide/#writing-tests-annotations the are more annotations, should we add them by default too ?

Default value for   property "annotation" will be "org.junit.jupiter.api.Test". Default value for "visibility" is "private". Am I right ?

Please extend your test class with more junit5 annotations. You will reuse such work later on, during UTs. All wording from description of issue will be part of documentation. You are NOT doing waste work now, all will be reused.

**Comment by Michael1993** [comment]
If the default value for the 'annotation(s)' parameter is `org.junit.jupiter.api.Test` then 'visibility' can not be `private` by default, unfortunately.

**Comment by romani** [comment]
Ok in this case let's remove from description of issue any reference to good design by closing access as possible. So Check rationale is simply keep code consistent. 

So, I propose to make default is public, as package-private is illusion of control over method calls. Test should not be reused so should never be called from other methods. In addition default scope is not good for human to recognize, so some best practices demand a comment instead of it, and we use such style.
As nuance, checkstyle have to use all Checks it host, so would prefer to have public scope as default if junit5 annotations are defaults.

**Comment by verhas** [comment]
>should we add them by default too ?

I recommend that the configuration should be `annotations` (plural) and only those JUnit 5 annotations should be added to the default array of annotations that denote that the annotated method is a test. For example `@DisplayName` alone does not signal that the annotated method is a test method.

The visibility default should be package-private as that is the minimum required visibility for tests.

"package-private is illusion of control over method calls"

I do not know what you mean by that. Could you elaborate?

"default scope is not good for human to recognize"

I tend to disagree.

"some best practices demand a comment instead of it"

Could you reference? I am interested who wrote that and in which book or where.

"As nuance, checkstyle have to use all Checks it host, so would prefer to have public scope as default if junit5 annotations are defaults."

Ha ha. Perhaps time to change the visibility of the test methods if CheckStyle uses JUnit 5. :-)

**Comment by romani** [comment]
>I recommend that the configuration should be annotations (plural) and only those JUnit 5 annotations

Let's make exact list of annotations.

>Could you reference? I am interested who wrote that and in which book or where.

https://github.com/checkstyle/checkstyle/issues/957

>Perhaps time to change the visibility of the test methods 

No. It was never a problem. In our case we will need to put comment instead as one rule is enforced. As it gives no benefits and security/control of execution. I do not see a reason. 

This property will be "public" by default, you can change it in your config.

**Comment by Michael1993** [comment]
If I understood the JUnit 5 documentation properly, the annotations in the issue description are the only annotations that denote a test. Other annotations denote dynamic test factories, templates, decorators, etc.

I have changed the issue description to use `public` for the `visibility` property.

@romani Please consider that many people treat checkstyle as a de-facto standard. Do you think my arguments for stricter default visibility are moot/irrelevant or...?

Another option is to **not** have a default value on the `visibility` property, instead requiring the user to define their own value.

**Comment by verhas** [comment]
Thanks for the reference. I have never met this practice as we never ever had any problem with package private members.  IntelliJ will warn while editing about the `public` keyword in JUnit 5 tests as it realizes that they do not need to be public and that IDE performs the code analysis as you edit the code.

Joshua Bloch Effective Java, Third Edition, Item#15 (Minimize the accessibility of classes and members) says that a member, like a method, should have the least possible visibility. In the case of JUnit 5 test methods, this visibility is package private. That is the reason why I thought that package-private would be the most sensible default.

The decision about the `public` or package-private as a default value in this situation also has a long term effect on the industry. Maybe it is not such a big issue as the null pointer in the 1960s , but checkstyle certainly has the momentum and is able to shape the industry. Millions of developers will ignore style guides, documentation or recommendations and will write the code that matches the default setup of their CI server that runs checkstyle. So if you decide that the default of this value is `public` then most of the developers will write public tests under JUnit 5. I actually do not know why they changed it from public to package-private when migrating from JUnit 4 to 5 but I am certain that they had some reason.

**Comment by romani** [comment]
>  Please consider that many people treat checkstyle as a de-facto standard.

probably, but I never recommending to blindly follow it in default form. Even we do not use all checks in default form. Human is still required to make a decision in what form to apply a Check to code base of certain team.

> Do you think my arguments for stricter default visibility are moot/irrelevant or...?

I appreciate your time and desire to raise it as an issue. I considered several options before make such decision:
- it is already "public" for two decades, now all will be happy to refactor all tests. But they could do this with extra change in config.
- there are other best-practices that enforce "explicit" definition of scope. I agree with explicit, but I do not agree that convertion from public to private-package will help to resolve/prevent issues. 
- This Check is NOT junit5. We will show user how it works on junit5 case, but it is not dedicated for junit5 best-practices. It is dedicated to keep code consistent in one and another form. Whole idea of Checkstyle is not participate in style wars and let different teams to enforce their understanding of right approach.

> It's easy to fall into the trap of thinking that 'it is not important

ones you send PR, you will understand that do follow this rule probably more than anybody in this world.
BUT "package-private" is sign of bad design FOR ME (subjective) even in main code. It is ok when you have no time to think about design, but it does not mean that it is good approach. The best design is when class is "non breakable" from any outside usage. Confusion is coming that classes are more aware of other classes internals in they are placed in same package !!!! IT IS NOT RIGHT. It is author/contributor of class/method is not aware of other class internals. So changing a location of class in package tree, does not extending of contributor of target class internals. Such practice of "friends"(c++) and "package-private"(java) come from time then there was not that much code and engineers were stick to code they owned to support, so it worked ok in such model. BUT in modern word non of engineers is glued to code, there are bunch of contributors can work/own on same code and good design do not let multiple owners (with different level of awarenes of inner details) misuse class usage.

> Another option is to not have a default value on the visibility property, instead requiring the user to define their own value.

no, it is limit of checkstyle design. Check should have default value/behavior.

> like a method, should have the least possible visibility

yes, but please do not be very fanatic with rule that is taken out of context. 
Please read my explanation in this comment post (a bit above).
It is hard to dispute .... most likely we will not come to same decision on this, special property will reconcile us. 

> Maybe it is not such a big issue as the null pointer in the 1960s

it is just another hysteria .... OK there are no NPE in code, and now they have NoSuchElementException and all other exceptions and badly working logic. In some cases it is the same.

> So if you decide that the default of this value is public then most of the developers will write public tests under JUnit 5

They will write code as before, consistent with previous huge amount of code. BUT if engineer is so like to finally fix this problem, he has ability to do this. In such case , this engineer is do care and will not follow defaults, and will not be lazy to update Check to work up to his expectations.
But see my reasons in second reply block in this post.

**Comment by romani** [comment]
@rnveach , please do final review of Check and approve issue.

**Comment by romani** [comment]
I have interesting idea:
This Check will target by default annotation `VisibleForTesting` from guava and validate that visibility of such methods are "protected" or "package-private".
I hope it is non disputable expectations from code, right ?

This gives additional insight in demand for this Check is to have array of visibility values (any match).

Are you ok with such default and amendments to design ?
In documentation examples, we can keep junit5 with suggestion to use package-private visibility.

**Comment by verhas** [comment]
I love your proposal. You are a great man.

**Comment by romani** [comment]
@Michael1993 , what do you think about proposed changes ?

**Comment by rnveach** [comment]
I am ok with design so far, but it sounds like we are still discussing changes.

**Comment by Michael1993** [comment]
I'm like your changes, I'll re-write the issue description tomorrow. Thank you for working with me on this and sorry for my late reply.

**Comment by romani** [comment]
@Michael1993 , ping

**Comment by Michael1993** [comment]
I rewrote the issue description, please check when you have some time @romani

**Comment by romani** [comment]
Thanks a lot, looks good. We might later on discuss list of all possible targets, but for start of implementation this set is enough.

@rnveach please finalize review and approval.

**Comment by gianmarcoschifone** [comment]
I'm on it

**Comment by gianmarcoschifone** [comment]
Hi @romani ,
Consider the following case:
```java
import x.x.x.Annotation;

public class Example {

    @y.y.y.Annotation
    void method1() {
    }

    @Annotation
    void method2() {
    }
}
```
If the check is configured to work only with `x.x.x.Annotation` or only with `y.y.y.Annotation`, simple-name matching would be ambiguous.

In this situation, should we resolve imports to determine the canonical annotation type, or would you prefer a simpler matching approach?

P.S. Since this is pretty old: no need to read all comments of the issue, the description is enough to understand the check

**Comment by romani** [comment]
By java rules, compilation, if import is defined, only this class can use short form of reference.
By scanning of imports , you can be confident in what you have in usage.

**Comment by gianmarcoschifone** [comment]
> By java rules, compilation, if import is defined, only this class can use short form of reference. By scanning of imports , you can be confident in what you have in usage.

Alright, thank you!

**Comment by romani** [comment]
fix is merged, will  be released in at the end of July
